system-tests/create_filewriter_image.py

```diff
@@ -0,0 +1,184 @@
+import os.path
+import os
+import pytest
+import docker
+import tarfile
+import hashlib
+import io
+
+IMAGE_NAME = "screamingudder/ubuntu18.04-build-node:3.0.6"
+TEST_IMAGE_NAME = "filewriter-image"
+CONTAINER_NAME = "filewriter-system-test"
+DEBUG_MODE = False
+client = docker.from_env()
+
+
+def execute_command(command, workdir, container):
+    exec_id = container.client.api.exec_create(container.id, command, workdir=workdir)[
+        "Id"
+    ]
+    output = container.client.api.exec_start(exec_id, stream=True)
+    for line in output:
+        if DEBUG_MODE:
+            print(line.decode("utf-8"))
+    exit_code = client.api.exec_inspect(exec_id)["ExitCode"]
+    if exit_code != 0:
+        raise RuntimeError(
+            "The following command failed with a non-zero exit code ({}): {}".format(
+                exit_code, command
+            )
+        )
+
+
+def copy_files_to_container(paths, base_path, container):
+    in_memory_file = io.BytesIO()
+    tar = tarfile.open(fileobj=in_memory_file, mode="w")
+    for path in paths:
+        tar.add(arcname=path, name=base_path + path)
+    tar.close()
+    container.put_archive("/home/jenkins/", in_memory_file.getvalue())
+    return hashlib.sha512(in_memory_file.getvalue()).hexdigest()
+
+
+def create_hash_file(file_name, hash):
+    out_file = open(file_name, "w")
+    out_file.write(hash)
+    out_file.close()
+
+
+def copy_to_container(container):
+    conanfile = "conan/conanfile.txt"
+    conan_hash = copy_files_to_container([conanfile,], "../", container)
+    source_files = ["cmake/", "src/", "CMakeLists.txt", "docker_launch.sh"]
+    src_hash = copy_files_to_container(source_files, "../", container)
+
+    src_hash_file_name = "src_hash_new.txt"
+    conan_hash_file_name = "conan_hash_new.txt"
+    create_hash_file(src_hash_file_name, src_hash)
+    create_hash_file(conan_hash_file_name, conan_hash)
+
+    copy_files_to_container([src_hash_file_name, conan_hash_file_name], "./", container)
+
+
+def generate_new_container():
+    environment_variables = {}
+
+    def add_var(var_name):
+        if var_name in os.environ:
+            environment_variables[var_name] = os.environ[var_name]
+
+    add_var("http_proxy")
+    add_var("https_proxy")
+    add_var("local_conan_server")
+    try:
+        client.images.get(IMAGE_NAME)
+    except docker.errors.ImageNotFound as e:
+        client.images.pull(IMAGE_NAME)
```

Comment:
Pulling the image can take a while, a print statement here would be good.

Reply:
👍